modules/xmpp/JingleSessionPC.js

```diff
@@ -503,25 +503,48 @@ export default class JingleSessionPC extends JingleSession {
             // IS_SELECTED_CHANGED events and notify the peer connection
             this.rtc.addListener(RTCEvents.IS_SELECTED_CHANGED,
                 isSelected => {
-                    this.peerconnection.setIsSelected(isSelected);
-                    logger.info('Doing local O/A due to '
-                        + 'IS_SELECTED_CHANGED event');
-                    this.modificationQueue.push(finishedCallback => {
-                        if (!this._assertNotEnded('setIsSelected O/A')) {
-                            finishedCallback();
-
-                            return;
-                        }
+                    this._executeOnTheQueue('setIsSelected', finishedCallback => {
+                        logger.info('Doing local O/A due to IS_SELECTED_CHANGED event');
+                        this.peerconnection.setIsSelected(isSelected);
 
                         this._renegotiate()
-                            .then(finishedCallback)
-                            .catch(finishedCallback);
+                            .then(finishedCallback, finishedCallback);
                     });
                 }
             );
         }
     }
 
+    /**
+     * Executes a task on the modification queue. See {@link _processQueueTasks} for more info.
+     *
+     * @param {string} actionName - Name of the action which will appear in the logs.
+     * @param {function} workFunction - The worker function, see {@link _processQueueTasks} for more info.
+     * @return {Promise<void>}
+     * @private
+     */
+    _executeOnTheQueue(actionName, workFunction) {
+        return new Promise((resolve, reject) => {
+            this.modificationQueue.push(
+                finishCallback => {
+                    if (!this._assertNotEnded(actionName)) {
+                        finishCallback();
+
+                        return;
+                    }
+
+                    workFunction(finishCallback);
```

Comment:
shouldn't there be a `resolve` somewhere in this block?

Reply:
EDIT: ok, i remember now why we don't see this. was talking with lenny and we thought maybe this would be clearer:

```
_executeOnTheQueue(actionName, workFunction) {
    return new Promise((resolve, reject) => {
        const wrappedWorkFunction = finishedCallback => {
            if (!this.assertNotEnded(actionName)) {
                finishedCallback();
                return;
            }
            workFunction(finishedCallback);
         }
         const handler = error => {
             if (error) { 
                 reject(error);
             } else {
                  resolve();
             }
          this.modificationQueue.push(wrappedWorkFunction, handler);
       });
}
```

i was hoping we could not wrap everything in the promise by doing it that way as well, but think we're stuck there.